### DISCUSSION TITLE
Importing pycardano throws OSError: could not get source code

### DISCUSSION THREAD

**Discussion by l-o-u-i-s-g** [beeware/toga#2704] [Q&A] [answered]
I'm trying to use a module that  installs successfully during the build but then fails at the import  when the app is started on android.  It works great on linux  so  what is different between android and linux? paths? 

More precisely I'm trying to use pycardano  in my project.  I added it to  my dependencies in  pyproject.toml 
Its getting installed (and all it dependencies) during the android build.
```
Collecting pycardano
  Using cached pycardano-0.11.1-py3-none-any.whl (83 kB)
Installing collected packages: typing-extensions, aiosqlite, chaquopy-libffi, pycparser, cffi, cryptography, six, argon2-cffi, mnemonic, certifi, idna, charset-normalizer, urllib3, requests, blockfrost-python, docker, ecdsa, attrs, cbor2, asn1crypto, oscrypto, certvalidator, cose, pprintpp, cachetools, frozendict, websocket-client, ECPy, zipp, importlib-metadata, typeguard, frozenlist, PyNaCl, pycardano, travertino, toga-core, toga-android
Successfully installed ECPy-1.2.5 PyNaCl-1.5.0 aiosqlite-0.20.0 argon2-cffi-20.1.0 asn1crypto-1.5.1 attrs-23.2.0 blockfrost-python-0.5.3 cachetools-5.3.3 cbor2-5.6.4 certifi-2024.7.4 certvalidator-0.11.1 cffi-1.13.2 chaquopy-libffi-3.3 charset-normalizer-3.3.2 cose-0.9.dev8 cryptography-3.4.8 docker-7.1.0 ecdsa-0.19.0 frozendict-2.4.4 frozenlist-1.4.1 idna-3.7 importlib-metadata-8.0.0 mnemonic-0.21 oscrypto-1.3.0 pprintpp-0.4.0 pycardano-0.11.1 pycparser-2.22 requests-2.32.3 six-1.16.0 toga-android-0.4.5 toga-core-0.4.5 travertino-0.3.0 typeguard-4.3.0 typing-extensions-4.12.2 urllib3-2.2.2 websocket-client-1.8.0 zipp-3.19.2
```
but when I run the build on my android device I get
```
E/AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.mydomain.myapp/org.beeware.android.MainActivity}: com.chaquo.python.PyException: OSError: could not get source code
E/AndroidRuntime: Caused by: com.chaquo.python.PyException: OSError: could not get source code
E/AndroidRuntime:       at <python>.inspect.findsource(inspect.py:798)
E/AndroidRuntime:       at <python>.inspect.getsourcelines(inspect.py:979)
E/AndroidRuntime:       at <python>.inspect.getsource(inspect.py:997)
E/AndroidRuntime:       at <python>.pycardano.types.typechecked(types.py:16)
E/AndroidRuntime:       at <python>.pycardano.serialization.<module>(serialization.py:218)
E/AndroidRuntime:       at <python>.pycardano.hash.<module>(hash.py:5)
E/AndroidRuntime:       at <python>.pycardano.address.<module>(address.py:20)
E/AndroidRuntime:       at <python>.pycardano.<module>(__init__.py:3)
```
for testing purposes  I commented my functions that requires the import and  the import itself.. then the app runs. Just uncommenting the import makes it crash.

**Comment by freakboy3742** [comment, marked as answer]
There's a *lot* of differences between Android and a desktop Linux - including, but not limited to, paths.

I have no idea what pycardarno is, so very difficult to comment on what exactly is going wrong here, beyond the detail that is already in the message - that something in the cardarno library is trying to access the source code of the module at runtime, and is doing that in a way that isn't compatible with deployment on Android.

This might be due to how code is deployed on Android. In a desktop deployment, an installed package is a directory of files on disk; however, on Android, the code is contained in the APK file, which is effectively a zip. When running on Android, a custom loader is used to import code; this works fine for "standard" Python imports, but if cordarno is relying on assumptions about the availability of code on disk that aren't actually guaranteed by Python's import system, then it could be causing problems.

A possible workaround is to use Briefcase's ability to extract files at runtime. This is done automatically for resource files like images, but you can force the whole of a source package to be unpacked at runtime using a configuration option in your `pyproject.toml` file. Adding something like:
```
build_gradle_extra_content = """
android.defaultConfig.python {
    extractPackages("pycardano")
}
"""
```
to the Android section of your `pyproject.toml` will add a directive to unpack the whole "pycardano" package at runtime. This will result in a slightly slower app startup on first run (because the files need to be unpacked), but the files will be available at runtime in a more traditional directory layout.

You may need to modify the specifics of this configuration for your purposes. I don't know what file exactly is being requested, what package it is in, or the package structure of pycordano; so a different `extractPackages()` invocation may be needed.

**Reply by l-o-u-i-s-g** [reply]
Thanks for your answer! Very well explained...
